### PR TITLE
Post-attention + post-MLP RMSNorm in TransformerBlock (sandwich-norm)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**Add post-attention and post-MLP RMSNorm** inside each Transformer block (sandwich-norm / "double-norm" pattern from Modded-NanoGPT and recent stable-attention work). The current `TransformerBlock` is plain pre-LN:

```python
x = x + self.attention(self.norm1(x))    # only pre-norm in front of attention
x = x + self.mlp(self.norm2(x))          # only pre-norm in front of MLP
```

Pre-LN-only Transformers are known to have growing residual-stream norms with depth, especially when the MLP / attention output is unbounded. Modded-NanoGPT showed that adding a post-block normalization on the **output** of each sublayer (before the residual add) noticeably stabilizes training and lets you push higher LR / longer schedules. The Transolver slot-attention is exactly the kind of unbounded-output sublayer that benefits — slot attention divides by softmaxed slot scores which can produce large activations on a few tokens.

This is a ~10-line code change with no new hyperparameters, ranked #2 in `research/CURRENT_RESEARCH_STATE.md` next-priorities.

## Instructions

Modify `target/model.py` `TransformerBlock` (line 177–204). Add two more norms — RMSNorm on the **output** of `self.attention(...)` and on the **output** of `self.mlp(...)`, before the residual add. Use RMSNorm rather than LayerNorm for the post norms to match the Modded-NanoGPT recipe (cheaper, no mean-subtract, which is what we want for outputs that should be near zero-mean already).

PyTorch 2.7 ships `torch.nn.RMSNorm`. Use it directly.

```python
# in TransformerBlock.__init__ (line 187 area):
mlp_hidden_dim = int(math.ceil(hidden_dim * mlp_expansion_factor))
self.norm1 = nn.LayerNorm(hidden_dim, eps=1e-6)
self.post_attn_norm = nn.RMSNorm(hidden_dim, eps=1e-6)   # NEW
self.attention = TransolverAttention(...)
self.norm2 = nn.LayerNorm(hidden_dim, eps=1e-6)
self.post_mlp_norm = nn.RMSNorm(hidden_dim, eps=1e-6)    # NEW
self.mlp = UpActDownMlp(...)
```

```python
# in TransformerBlock.forward (line 198):
def forward(self, x, attn_mask=None):
    x = _apply_token_mask(x, attn_mask)
    x = x + self.post_attn_norm(self.attention(self.norm1(x), attn_mask=attn_mask))
    x = _apply_token_mask(x, attn_mask)
    x = x + self.post_mlp_norm(self.mlp(self.norm2(x)))
    x = _apply_token_mask(x, attn_mask)
    return x
```

That is the entire structural change. Do not change any of:
- the optimizer / LR / WD / EMA / batch / point counts
- the model heads / layers / hidden_dim / slices / mlp_ratio
- gradient clipping, warmup, T_max

This is a pure architectural ablation against PR #232 SOTA. Anything else changing makes attribution impossible.

## Run plan — single 8-GPU run, full SOTA stack

```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent tanjiro --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --ema-decay 0.999 --lr-warmup-epochs 1 \
  --wandb_group post-attn-rmsnorm
```

If the first run beats SOTA val_abupt < 9.065 % at any epoch, you're done — submit for review. If it's within ±0.5pp at ep11, also submit (we want post-norm runs even if neutral, to combine with Muon later). If it diverges or trains substantially worse (val > 12% by ep5), file a comment with the loss curve and we'll close.

## Sanity checks

- Print the new model parameter count at startup; the increase from 2 RMSNorm × num_layers × hidden_dim should be tiny (e.g. for 4L × 512d that's only +4096 params).
- Spot-check that `RMSNorm` is being applied *after* `self.attention(...)` and `self.mlp(...)`, not in front (the pre-norms `norm1` / `norm2` must remain).

## Baseline (PR #232 — model_heads=4 SOTA)

| Metric | Value |
|---|---|
| **val_abupt (best)** | **9.0650 %** (ep11) |
| **test_abupt** | **10.190 %** |
| W&B run | `r8s2dtnq` (`wandb-applied-ai-team/senpai-v1-drivaerml-ddp8`) |

Per-axis test:
- surface_pressure 5.461 % | wall_shear 9.910 % | volume_pressure 12.656 %
- tau_x 8.432 % | tau_y 11.952 % | tau_z 12.447 %

AB-UPT references (target gaps): surface_pressure 3.82 | wall_shear 7.29 | volume_pressure 6.08 | tau_x 5.35 | tau_y 3.65 | tau_z 3.63.

**Reproduce SOTA (no architecture change):**
```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent <STUDENT> --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --ema-decay 0.999 --lr-warmup-epochs 1
```

## Reporting

In the PR results comment include:
- Best val_abupt across all epochs and the epoch index.
- Best-checkpoint test_abupt + per-axis breakdown.
- W&B run ID.
- Any observation about residual-stream norm stability vs SOTA (e.g. did training noise drop?). This informs whether to also try LR-up follow-ups.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #300 because ADVISOR: NEGATIVE result. Sandwich-norm (post-attn + post-MLP RMSNorm) caused catastrophic divergence. Run 528uuqx5 reached val_primary/abupt_axis_mean_rel_l2_pct=79.98% (vs baseline 9.065%) — model is failing to learn. Per-axis: surface_pressure 67.5%, volume_pressure 62.3%, wall_shear 83.97%, tau_y 99.27%, tau_z 95.06%. The double-norm hypothesis from Modded-NanoGPT does not transfer to Transolver slot attention — likely because the slot routing already produces near-RMS-1 features and the additional post-norms collapse signal. Earlier runs in same group also crashed (40t1yerr at 66.18%). Closing — tanjiro will be reassigned a fresh experiment..
